flytekit/core/data_persistence.py

```diff
@@ -0,0 +1,382 @@
+        # Ignore add_protocol for now. Only complicates things
+        if add_prefix:
+            return os.path.join(self.default_prefix, *args)
+        return os.path.join(*args)
+
+
+class FileAccessProvider(object):
+    """
+    This is the class that is available through the FlyteContext and can be used for persisting data to the remote
+    durable store.
+    """
+
+    def __init__(self, local_sandbox_dir: Union[str, os.PathLike], raw_output_prefix: str):
+        # Local access
+        if local_sandbox_dir is None or local_sandbox_dir == "":
+            raise Exception("Can't use empty path")
+        local_sandbox_dir_appended = os.path.join(local_sandbox_dir, "local_flytekit")
+        self._local_sandbox_dir = pathlib.Path(local_sandbox_dir_appended)
+        self._local_sandbox_dir.mkdir(parents=True, exist_ok=True)
+        self._local = DiskPersistence(default_prefix=local_sandbox_dir_appended)
+
+        self._default_remote = DataPersistencePlugins.find_plugin(raw_output_prefix)(default_prefix=raw_output_prefix)
+        self._raw_output_prefix = raw_output_prefix
+
+    @staticmethod
+    def is_remote(path: Union[str, os.PathLike]) -> bool:
+        """
+        Deprecated. Lets find a replacement
+        """
+        if path.startswith("/") or path.startswith("file://"):
+            return False
+        return True
+
+    @property
+    def local_sandbox_dir(self) -> os.PathLike:
+        return self._local_sandbox_dir
+
+    @property
+    def local_access(self) -> DiskPersistence:
+        return self._local
+
+    def construct_random_path(
+        self, persist: DataPersistence, file_path_or_file_name: typing.Optional[str] = None
+    ) -> str:
+        """
+        Use file_path_or_file_name, when you want a random directory, but want to preserve the leaf file name
+        """
+        key = UUID(int=random.getrandbits(128)).hex
+        if file_path_or_file_name:
+            _, tail = os.path.split(file_path_or_file_name)
+            if tail:
+                return persist.construct_path(False, True, key, tail)
+            else:
+                logger.warning(f"No filename detected in {file_path_or_file_name}, generating random path")
+        return persist.construct_path(False, True, key)
+
+    def get_random_remote_path(self, file_path_or_file_name: typing.Optional[str] = None) -> str:
+        """
+        Constructs a randomized path on the configured raw_output_prefix (persistence layer). the random bit is a UUID
+        and allows for disambiguating paths within the same directory.
+
+        Use file_path_or_file_name, when you want a random directory, but want to preserve the leaf file name
+        """
+        return self.construct_random_path(self._default_remote, file_path_or_file_name)
+
+    def get_random_remote_directory(self):
+        return self.get_random_remote_path(None)
+
+    def get_random_local_path(self, file_path_or_file_name: typing.Optional[str] = None) -> str:
```

Comment:
can file_path_or_file_name be `os.PathLIke`?

Reply:
not necessarily. Its actually a string